ocaml/parsing/jane_syntax.ml

```diff
@@ -618,6 +623,15 @@ module Jkind = struct
       bind (struct_item_to_type item_of_ty) (fun ty -> ret loc (Kind_of ty))
     | Some ("abbrev", [item], loc) ->
       bind (Const.of_structure_item item) (fun c -> ret loc (Abbreviation c))
+    | Some ("arrow", result :: args, loc) ->
+      bind (of_structure_item result) (fun { txt = result } ->
+          let args = List.map of_structure_item args in
+          if List.for_all Option.is_some args
+          then
+            let args = List.map Option.get args in
+            let args = List.map (fun { txt } -> txt) args in
+            ret loc (Arrow (args, result))
+          else None)
```

Comment:
It's best to avoid `Option.get` where you can. Here, you can instead do:
```ocaml
let result = of_structure_item result in
let args = 
  List.fold_right
     (fun arg acc ->
        bind acc (fun acc ->
          bind (of_structure_item arg) (fun arg -> arg :: acc)))
     args (Some [])
in
bind result (fun { txt = result } ->
  bind args (fun args ->
     ret loc (Arrow (args, result))))
```
You lose the short-circuiting but that only happens in the error case anyway, and now it is clearer that you are just accumulating arguments in the same way as the surrounding code. (If this function used binding operators, this monadic style would look nicer, but it's all going to be deleted soon so no point improving it now).

I'd probably lift the `List.fold_right` part out to be mutually recursive with `of_structure_item` and call it something like `of_structure_item_list` or maybe `list_of_structure_item_list`. So that the main body becomes just:
```ocaml
bind (of_structure_item result) (fun { txt = result } ->
  bind (list_of_structure_item_list args) (fun args ->
     ret loc (Arrow (args, result))))
```

Reply:
It'd be nice to have `all : 'a option list -> 'a list option` somewhere, as the nested monadic style isn't particularly easy to read for me.

Reply:
You could put it in `Misc.Stdlib.Option`, but I think it's probably better not to in this case. The name `all` is a bit of a Jane St.-ism and isn't used elsewhere in the compiler. Since this code is on the way out anyway I'd rather not add utility functions for it.